File: api/SdCard.go
package api

import (
	"time"

	"github.com/lvcdy/huawei-lte-api-go/session"
)

// SdCard 对应 SdCard.py。
type SdCard struct {
	*session.ApiGroup
}

// NewSdCard 创建 SdCard API 分组。
func NewSdCard(s *session.Session) *SdCard {
	return &SdCard{ApiGroup: session.NewApiGroup(s)}
}

// DlnaSetting 对应 sdcard/dlna-setting。
func (s *SdCard) DlnaSetting() (map[string]interface{}, error) {
	return s.S.Get("sdcard/dlna-setting", nil, "api")
}

// SetDlnaSetting 设置 DLNA。对应 sdcard/dlna-setting (post_set)。
func (s *SdCard) SetDlnaSetting(enabled bool, shareAll bool, sharePath string) (interface{}, error) {
	return s.S.PostSet("sdcard/dlna-setting", map[string]interface{}{
		"enabled":      boolToInt(enabled),
		"sharepath":    sharePath,
		"shareallpath": boolToInt(shareAll),
	}, false, "api", false, false)
}

// Sdcard 对应 sdcard/sdcard。
func (s *SdCard) Sdcard() (map[string]interface{}, error) {
	return s.S.Get("sdcard/sdcard", nil, "api")
}

// Sdcardsamba 对应 sdcard/sdcardsamba。
func (s *SdCard) Sdcardsamba() (map[string]interface{}, error) {
	return s.S.Get("sdcard/sdcardsamba", nil, "api")
}

// SetSdcardsamba 使用 SMB 启用文件共享。对应 sdcard/sdcardsamba (post_set)。
func (s *SdCard) SetSdcardsamba(
	enabled bool,
	serverName string,
	serverDescription string,
	workgroupName string,
	anonymousAccess bool,
	printerEnabled bool,
) (interface{}, error) {
	return s.S.PostSet("sdcard/sdcardsamba", session.O(
		"enabled", boolToInt(enabled),
		"servername", serverName,
		"serverdescription", serverDescription,
		"workgroupname", workgroupName,
		"anonymousaccess", boolToInt(anonymousAccess),
		"printerenable", boolToInt(printerEnabled),
	), false, "api", false, false)
}

// Printerlist 对应 sdcard/printerlist。
func (s *SdCard) Printerlist() (map[string]interface{}, error) {
	return s.S.Get("sdcard/printerlist", nil, "api")
}

// ShareAccount 对应 sdcard/share-account。
func (s *SdCard) ShareAccount() (map[string]interface{}, error) {
	return s.S.Get("sdcard/share-account", nil, "api")
}

// Sdfile 对应 sdcard/sdfile。
func (s *SdCard) Sdfile() (map[string]interface{}, error) {
	return s.S.Get("sdcard/sdfile", nil, "api")
}

// Fileupload 对应 sdcard/fileupload。
func (s *SdCard) Fileupload() (map[string]interface{}, error) {
	return s.S.Get("sdcard/fileupload", nil, "api")
}

// CheckFileExist 对应 sdcard/Check_file_exist。
func (s *SdCard) CheckFileExist() (map[string]interface{}, error) {
	return s.S.Get("sdcard/Check_file_exist", nil, "api")
}

// CreateDir 在 SD 卡上创建目录。对应 sdcard/createdir (post_set)。
func (s *SdCard) CreateDir(name string, currentPath string, created *time.Time) (interface{}, error) {
	createTime := time.Now()
	if created != nil {
		createTime = *created
	}
	return s.S.PostSet("sdcard/createdir", map[string]interface{}{
		"CurrentPath": currentPath,
		"FileName":    name,
		"Time": map[string]interface{}{
			"Year":  createTime.Year(),
			"Month": int(createTime.Month()),
			"Day":   createTime.Day(),
			"Hour":  createTime.Hour(),
			"Min":   createTime.Minute(),
			"Sec":   createTime.Second(),
		},
	}, false, "api", false, false)
}

// DeleteFile 删除 SD 卡上的文件/目录。对应 sdcard/deletefile (post_set)。
func (s *SdCard) DeleteFile(name string, currentPath string) (interface{}, error) {
	return s.S.PostSet("sdcard/deletefile", map[string]interface{}{
		"CurrentPath":    currentPath,
		"DeleteFileList": name,
	}, false, "api", false, false)
}

// SdCapacity 获取 SD 卡容量信息。对应 sdcard/sdcapacity。
func (s *SdCard) SdCapacity() (map[string]interface{}, error) {
	return s.S.Get("sdcard/sdcapacity", nil, "api")
}

File: src/api/sd_card.rs
//! SdCard API group (`api/SdCard.py`).
//!
//! Storage sharing (DLNA / Samba) and SD-card file management.

use chrono::{Datelike, Timelike};
use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::{XmlMap, XmlValue};

use super::{get_value, post_set};

/// SdCard API group.
pub struct SdCard<'a> {
    conn: &'a Connection,
}

impl<'a> SdCard<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        SdCard { conn }
    }

    /// `sdcard/dlna-setting`. DLNA setting.
    pub fn dlna_setting(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/dlna-setting")
    }

    /// Set DLNA settings.
    ///
    /// * `enabled` — enable DLNA.
    /// * `share_all` — share the whole card.
    /// * `share_path` — path to share (default `/`).
    pub fn set_dlna_setting(
        &self,
        enabled: bool,
        share_all: bool,
        share_path: &str,
    ) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "enabled",
                if enabled {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("sharepath", share_path.to_string()),
            (
                "shareallpath",
                if share_all {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
        ]);
        post_set(self.conn, "api/sdcard/dlna-setting", &body)
    }

    /// `sdcard/sdcard`. Sharing information.
    pub fn sdcard(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/sdcard")
    }

    /// `sdcard/sdcardsamba`. Samba sharing state.
    pub fn sdcardsamba(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/sdcardsamba")
    }

    /// Enable file sharing over SMB.
    pub fn set_sdcardsamba(
        &self,
        enabled: bool,
        server_name: &str,
        server_description: &str,
        workgroup_name: &str,
        anonymous_access: bool,
        printer_enabled: bool,
    ) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "enabled",
                if enabled {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("servername", server_name.to_string()),
            ("serverdescription", server_description.to_string()),
            ("workgroupname", workgroup_name.to_string()),
            (
                "anonymousaccess",
                if anonymous_access {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            (
                "printerenable",
                if printer_enabled {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
        ]);
        post_set(self.conn, "api/sdcard/sdcardsamba", &body)
    }

    /// `sdcard/printerlist`. Printer list.
    pub fn printerlist(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/printerlist")
    }

    /// `sdcard/share-account`. Share accounts.
    pub fn share_account(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/share-account")
    }

    /// `sdcard/sdfile` (reverse engineered).
    pub fn sdfile(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/sdfile")
    }

    /// `sdcard/fileupload` (reverse engineered).
    pub fn fileupload(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/fileupload")
    }

    /// `sdcard/Check_file_exist` (reverse engineered; note the mixed case).
    pub fn check_file_exist(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/Check_file_exist")
    }

    /// Create a directory on the SD card.
    ///
    /// * `name` — directory name to create.
    /// * `current_path` — parent path (default `/`).
    /// * `created` — creation timestamp (defaults to now in UTC).
    pub fn create_dir(
        &self,
        name: &str,
        current_path: &str,
        created: Option<chrono::DateTime<chrono::Utc>>,
    ) -> Result<String> {
        let created = created.unwrap_or_else(chrono::Utc::now);
        let time: XmlMap = map_of([
            ("Year", created.year().to_string()),
            ("Month", created.month().to_string()),
            ("Day", created.day().to_string()),
            ("Hour", created.hour().to_string()),
            ("Min", created.minute().to_string()),
            ("Sec", created.second().to_string()),
        ]);
        let mut body = XmlMap::new();
        body.insert(
            "CurrentPath".to_string(),
            XmlValue::Text(current_path.to_string()),
        );
        body.insert("FileName".to_string(), XmlValue::Text(name.to_string()));
        body.insert("Time".to_string(), XmlValue::Map(time));
        post_set(self.conn, "api/sdcard/createdir", &body)
    }

    /// Delete a file or directory on the SD card.
    pub fn delete_file(&self, name: &str, current_path: &str) -> Result<String> {
        let body: XmlMap = map_of([
            ("CurrentPath", current_path.to_string()),
            ("DeleteFileList", name.to_string()),
        ]);
        post_set(self.conn, "api/sdcard/deletefile", &body)
    }

    /// `sdcard/sdcapacity`. SD-card capacity information.
    pub fn sd_capacity(&self) -> Result<Value> {
        get_value(self.conn, "api/sdcard/sdcapacity")
    }
}
